pkg/alibabacloud/api/api.go

```diff
@@ -0,0 +1,501 @@
+	_, err := c.ecsClient.UnassignPrivateIpAddresses(req)
+	return err
+}
+
+func (c *Client) describeNetworkInterfaces(ctx context.Context, subnets ipamTypes.SubnetMap) ([]ecs.NetworkInterfaceSet, error) {
+	var result []ecs.NetworkInterfaceSet
+
+	for _, subnet := range subnets {
+		for i := 1; ; {
+			req := ecs.CreateDescribeNetworkInterfacesRequest()
+			req.PageNumber = requests.NewInteger(i)
+			req.PageSize = requests.NewInteger(50)
+			req.VSwitchId = subnet.ID
+			c.limiter.Limit(ctx, "DescribeNetworkInterfaces")
+			resp, err := c.ecsClient.DescribeNetworkInterfaces(req)
+			if err != nil {
+				return nil, err
+			}
+			if len(resp.NetworkInterfaceSets.NetworkInterfaceSet) == 0 {
+				break
+			}
+
+			for _, v := range resp.NetworkInterfaceSets.NetworkInterfaceSet {
+				result = append(result, v)
+			}
+			if resp.TotalCount < resp.PageNumber*resp.PageSize {
+				break
+			}
+			i++
+		}
+	}
+
+	return result, nil
+}
+
+// deriveStatus returns a status string based on the HTTP response provided by
+// the AlibabaCloud API server. If no specific status is provided, either "OK" or
+// "Failed" is returned based on the error variable.
+func deriveStatus(err error) string {
+	var respErr httperr.Error
+	if errors.As(err, &respErr) {
+		return respErr.ErrorCode()
+	}
+
+	if err != nil {
+		return "Failed"
+	}
+
+	return "OK"
+}
+
+// parseENI parses a ecs.NetworkInterface as returned by the ecs service API,
+// converts it into a eniTypes.ENI object
+func parseENI(iface *ecs.NetworkInterfaceSet, vpcs ipamTypes.VirtualNetworkMap, subnets ipamTypes.SubnetMap) (instanceID string, eni *eniTypes.ENI, err error) {
+	var privateIPSets []eniTypes.PrivateIPSet
+	for _, p := range iface.PrivateIpSets.PrivateIpSet {
+		privateIPSets = append(privateIPSets, eniTypes.PrivateIPSet{
+			Primary:          p.Primary,
+			PrivateIpAddress: p.PrivateIpAddress,
+		})
+	}
+
+	eni = &eniTypes.ENI{
+		NetworkInterfaceID: iface.NetworkInterfaceId,
+		MACAddress:         iface.MacAddress,
+		Type:               iface.Type,
+		InstanceID:         iface.InstanceId,
+		SecurityGroupIDs:   iface.SecurityGroupIds.SecurityGroupId,
+		VPC: eniTypes.VPC{
+			VPCID: iface.VpcId,
+		},
+		ZoneID: iface.ZoneId,
+		VSwitch: eniTypes.VSwitch{
+			VSwitchID: iface.VSwitchId,
+		},
+		PrimaryIPAddress: iface.PrivateIpAddress,
+		PrivateIPSets:    privateIPSets,
+		Tags:             parseECSTags(iface.Tags.Tag),
+	}
+	vpc, ok := vpcs[iface.VpcId]
+	if ok {
+		eni.VPC.CIDRBlock = vpc.PrimaryCIDR
+	}
+
+	subnet, ok := subnets[iface.VSwitchId]
+	if ok && subnet.CIDR != nil {
+		eni.VSwitch.CIDRBlock = subnet.CIDR.String()
+	}
+	return iface.InstanceId, eni, nil
+}
+
+// parseECSTags convert ECS Tags to ipam Tags
+func parseECSTags(tags []ecs.Tag) ipamTypes.Tags {
+	result := make(ipamTypes.Tags, len(tags))
+	for _, tag := range tags {
+		result[tag.TagKey] = tag.TagValue
+	}
+	return result
+}
+
+func min(a, b int) int {
+	if a > b {
+		return b
+	}
+	return a
+}
```

Comment:
There's probably no need for this. The `math` package has an implementation for `min`, let's use that.